README.md
# MMM-MyQ

[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](https://raw.githubusercontent.com/parnic/MMM-MyQ/main/LICENSE)

MyQ module for MagicMirror². Enables display and control of MyQ devices (specifically garage door openers) from Chamberlain, LiftMaster.

Currently only supports garage doors. Support for lights/other MyQ devices is possible, but not currently planned.

## Dependencies

* An installation of [MagicMirror²](https://github.com/MichMich/MagicMirror)
* npm

## Installation

1. Clone this repo into `MagicMirror/modules` directory.
1. Configure your `MagicMirror/config/config.js`:

    ```js
    {
        module: 'MMM-MyQ',
        position: 'top_right',
        config: {
            email: 'you@example.com',
            password: 'correct horse battery staple'
        }
    }
    ```

1. Run command `npm install --production` in your `modules/MMM-MyQ` directory.

## Config Options

| **Option** | **Default** | **Description** |
| --- | --- | --- |
| `email` | `''` | Your MyQ login address. (required) |
| `password` | `''` | Your MyQ login password. (required) |
| `types` | `['wifigaragedooropener']` | The types of MyQ devices to track. See [here](https://github.com/parnic/myq-api/blob/develop/src/constants.js#L8-L11) for a list of known device types. |
| `updateInterval` | `5 * 60 * 1000` (5 minutes) | How often to update the state of your devices, in milliseconds. |


CHANGELOG.md
# MMM-MyQ Changelog

## [1.0.2] - 2021-01-19

### Fixed

- Updated MyQ API to fix SERVICE_UNREACHABLE errors. This change requires `npm install` to be run again after updating.

## [1.0.1] - 2020-12-24

### Fixed

- Updated MyQ API to fix SERVICE_UNREACHABLE errors. This change requires `npm install` to be run again after updating.

## [1.0.0]

- Initial version


node_helper.js
/* Magic Mirror
 * Module: MMM-MyQ
 *
 * By parnic https://github.com/parnic/MMM-MyQ
 * MIT Licensed.
 */

const NodeHelper = require('node_helper');
const MyQ = require('myq-api');

module.exports = NodeHelper.create({
    _customActions: {
        open: 'open',
        close: 'close'
    },

    start() {
        console.log(`Starting module helper: ${this.name}`);
    },

    resetUpdates() {
        if (this.intervalId) {
            clearInterval(this.intervalId);
        }

        this.intervalId = setInterval(() => {
            this.getData();
        }, this.config.updateInterval);
    },

    socketNotificationReceived(notification, payload) {
        if (notification === 'MYQ_CONFIG') {
            this.config = payload;
            this.account = new MyQ();
            this.login();
        } else if (this.config) {
            if (notification === 'MYQ_TOGGLE') {
                const {device, action} = payload;
                const myqAction = action === this._customActions.open ? MyQ.actions.door.OPEN : MyQ.actions.door.CLOSE;

                this.account.setDoorState(device.serial_number, myqAction)
                    .then((result) => {
                        this.sendSocketNotification('MYQ_TOGGLE_COMPLETE', result.code === 0);
                    })
                    .catch((error) => this.handleError(error, 'setDoorState', true));
            } else if (notification === 'MYQ_UPDATE') {
                this.getData();
            }
        }
    },

    getData() {
        this.resetUpdates();

        this.account.getDevices()
            .then((result) => {
                if (result.code !== MyQ.constants.codes.OK) {
                    this.handleError(result, 'getDevices - then', true);
                    return;
                }

                let devices = result.devices.filter((device) => this.config.types.includes(device.device_type));
                this.sendSocketNotification('MYQ_DEVICES_FOUND', devices);
            })
            .catch((error) => this.handleError(error, 'getDevices', true));
    },

    login() {
        this.account.login(this.config.email, this.config.password)
            .then((result) => {
                if (result.code !== MyQ.constants.codes.OK) {
                    this.handleError(result, 'login - then');
                    return;
                }

                this.sendSocketNotification('MYQ_LOGGED_IN', {constants: MyQ.constants, actions: this._customActions});
                this.getData();
            })
            .catch((error) => this.handleError(error, 'login'));
    },

    handleError(error, context, attemptLogin) {
        if (attemptLogin && error.code === MyQ.constants.codes.LOGIN_REQUIRED) {
            this.login();
        } else {
            let err = `${context} failed with code ${error.code}`;
            this.sendSocketNotification('MYQ_ERROR', {context, err});
            console.error(`${this.name} ${context} error: ${err}`);
        }
    }
});


MMM-MyQ.js
/* Magic Mirror
 * Module: MMM-MyQ
 *
 * By parnic https://github.com/parnic/MMM-MyQ
 * MIT Licensed.
 */

Module.register('MMM-MyQ', {
    defaults: {
        email: '',
        password: '',
        types: ['wifigaragedooropener'],
        updateInterval: 5 * 60 * 1000 // every 5 minutes
    },

    states: {
        closed: 'closed',
        open: 'open',
        closing: 'closing',
        opening: 'opening'
    },

    getStyles() {
        return ['MMM-MyQ.css'];
    },

    start() {
        Log.info(`Starting module: ${this.name}`);
        this.devices = [];
        this.sendSocketNotification('MYQ_CONFIG', this.config);
    },

    socketNotificationReceived(notification, payload) {
        if (notification === 'MYQ_LOGGED_IN') {
            let {constants, actions} = payload;
            this.constants = constants;
            this.actions = actions;
        } else if (notification === 'MYQ_ERROR') {
            const {context, err} = payload;
            Log.error(`context=${context}, err=${err.message}`);
        } else if (notification === 'MYQ_TOGGLE_COMPLETE') {
            this.scheduleUpdate();
        } else if (notification === 'MYQ_DEVICES_FOUND') {
            payload.forEach((entry) => {
                let existingDevice = this.getDeviceBySerial(entry.serial_number);
                if (!existingDevice) {
                    this.devices.push(entry);
                    existingDevice = entry;
                } else {
                    existingDevice.state = entry.state;
                }

                if (existingDevice.desiredState) {
                    if (existingDevice.state.door_state !== existingDevice.desiredState) {
                        this.scheduleUpdate();
                    } else {
                        existingDevice.desiredState = null;
                    }
                }
            });

            this.updateDom();
        }
    },

    scheduleUpdate() {
        if (!this.timeout) {
            this.timeout = setTimeout(() => {
                this.timeout = null;
                this.sendSocketNotification('MYQ_UPDATE');
            }, 5000);
        }
    },

    getDom() {
        const wrapper = document.createElement('div');

        this.devices.forEach((device) => { wrapper.appendChild(this.getDeviceDom(device)); });

        return wrapper;
    },

    getDeviceDom(device) {
        const nameLabel = document.createElement('div');
        nameLabel.textContent = device.name;

        const btn = document.createElement('button');
        btn.appendChild(nameLabel);

        if (device.desiredState) {
            btn.disabled = true;
        } else {
            const actionLabel = document.createElement('div');
            let action = this.actions.close;
            actionLabel.textContent = this.translate('Opened');
            if (device.state.door_state === this.states.closed) {
                action = this.actions.open;
                actionLabel.textContent = this.translate('Closed');
            }

            btn.dataset['action'] = action;
            btn.appendChild(actionLabel);
        }

        btn.dataset['serial_number'] = device.serial_number;
        btn.classList.add('control', 'light');
        btn.addEventListener('click', (evt) => { this.deviceBtnAction(btn); });

        return btn;
    },

    deviceBtnAction(btn) {
        let device = this.getDeviceBySerial(btn.dataset['serial_number']);
        let action = btn.dataset['action'];

        if (device && action) {
            device.desiredState = action === this.actions.close ? this.states.closed : this.states.open;
            this.sendSocketNotification('MYQ_TOGGLE', {device, action});
            this.updateDom();
        }
    },

    getDeviceBySerial(serialNumber) {
        return this.devices.find((device) => device.serial_number === serialNumber);
    }
});
